doc/floppycontrol.1
.TH floppycontrol 1 "03Mar05" fdutils-5.5
.SH Name
floppycontrol - floppy driver configuration utility
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note
This manpage has been automatically generated from fdutils's texinfo
documentation.  However, this process is only approximative, and some
items, such as cross-references, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, only the items specific
to each command have been translated, and the general information about
fdutils has been dropped in the manpage version.  Thus I strongly advise
you to use the original texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips fdutils.dvi
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.TP
* \ \ 
To generate a HTML copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.lp
 
\&\fRA pre-made HTML can be found at:
\&\fR\&\f(CW\(ifhttp://www.tux.org/pub/knaff/fdutils\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as HTML.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.SH Description
.iX "p floppycontrol"
.iX "c configuration of floppy driver"
.PP
 
.nf
.ft 3
.in +0.3i
\&\fR\&\f(CWfloppycontrol [\fR\&\f(CW-p] [\fR\&\f(CW--pollstate] [\fR\&\f(CW--printfdstate]
[\fR\&\f(CW-a \fIoperation-abort-threshold\fR\&\f(CW] [\fR\&\f(CW-c \fIread-track-threshold\fR\&\f(CW]
[\fR\&\f(CW-r \fIrecalibrate-threshold\fR\&\f(CW] [\fR\&\f(CW-R \fIreset-threshold\fR\&\f(CW]
[\fR\&\f(CW-e \fIreporting-threshold\fR\&\f(CW] [\fR\&\f(CW-f] [\fR\&\f(CW-x] [\fR\&\f(CW-d \fIdrive\fR\&\f(CW][\fR\&\f(CW-F] [\fR\&\f(CW-T]
[\fR\&\f(CW-reset \fIcondition\fR\&\f(CW] [\fR\&\f(CW--debug] [\fR\&\f(CW--nodebug] [\fR\&\f(CW--messages]
[\fR\&\f(CW--nomessages] [\fR\&\f(CW--broken_dcl] [\fR\&\f(CW--working_dcl] [\fR\&\f(CW--inverted_dcl]
[\fR\&\f(CW--no_inverted_dcl] [\fR\&\f(CW--silent_dcl_clear] [\fR\&\f(CW--noisy_dcl_clear]
[\fR\&\f(CW-c\fIcmos-type\fR\&\f(CW] [\fR\&\f(CW-hlt \fIhlt\fR\&\f(CW] [\fR\&\f(CW-hut \fIhut\fR\&\f(CW] [\fR\&\f(CW-srt \fIsrt\fR\&\f(CW] [\fR\&\f(CW-o \fIspindown\fR\&\f(CW]
[\fR\&\f(CW-u \fIspinup\fR\&\f(CW] [\fR\&\f(CW-s \fIselect-delay\fR\&\f(CW] [\fR\&\f(CW-rps \fIrotations-per-second\fR\&\f(CW]
[\fR\&\f(CW-O \fIspindown-offset\fR\&\f(CW] [\fR\&\f(CW-track \fImax-tracks\fR\&\f(CW] [\fR\&\f(CW-timeout \fIseconds\fR\&\f(CW]
[\fR\&\f(CW-C \fIcheck-interval\fR\&\f(CW] [\fR\&\f(CW-n \fInative-format\fR\&\f(CW]
[\fR\&\f(CW-autodetect \fIautodetection-sequence\fR\&\f(CW] [\fR\&\f(CW-P] [\fR\&\f(CW--clrwerror]
[\fR\&\f(CW--printwerror] [\fR\&\f(CW-h]
.fi
.in -0.3i
.ft R
.lp
 
\&\fR
.PP
The \fR\&\f(CWfloppycontrol\fR program is used to configure the floppy driver.
.PP
.SH General\ Options
.IP
.TP
\&\fR\&\f(CW-h\fR\ 
.TQ
\&\fR\&\f(CW--help\fR
Print a help screen.
.TP
\&\fR\&\f(CW-d\ \fIdrive\fR\&\f(CW\ \fR\ 
.TQ
\&\fR\&\f(CW--drive\ \fI\ drive\fR\&\f(CW\fR
Selects the drive to configure. The default is drive 0
(\fR\&\f(CW\(if/dev/fd0\(is\fR).
.PP
.SH One\ time\ actions
.iX "c ejecting a disk"
.iX "c flushing floppy cache"
.iX "c resetting controller"
.PP
The following \fR\&\f(CWfloppycontrol\fR options don't set a configuration
parameter, but perform a one-time action. They are available to anybody
who has write access to the drive
.TP
\&\fR\&\f(CW-f\ \ \fR\ 
.TQ
\&\fR\&\f(CW--flush\fR
Flushes (throws away) the dirty data buffers associated with this drive.
.TP
\&\fR\&\f(CW-x\ \ \fR\ 
.TQ
\&\fR\&\f(CW--eject\fR
Ejects the disk out of the drive (Sparc). The dirty buffers are first
committed to disk before ejecting it. Fails if the disk is mounted.
.TP
\&\fR\&\f(CW--reset\ \fI\ condition\fR\&\f(CW\fR\ 
Resets the FDC under
\&\fIcondition\fR . Condition may be one of the following:
.RS
.TP
\&\fR\&\f(CW0\fR\ 
resets the FDC only if a reset is needed anyways,
.TP
\&\fR\&\f(CW1\fR\ 
resets the FDC
also if a raw command has been performed since the last reset, and 
.TP
\&\fR\&\f(CW2\ \fR\ 
resets the FDC unconditionally.
.RE
.IP
This command may be needed after some failed raw commands
(see section  fdrawcmd).
.IP
.TP
\&\fR\&\f(CW-F\fR\ 
.TQ
\&\fR\&\f(CW--formatend\fR
Issues an end format ioctl. This might be needed after exiting a
\&\fR\&\f(CWfdformat\fR in an unclean way. \fR\&\f(CWsuperformat\fR is not subject to
this.
.PP
.SH Printing\ current\ settings
.iX "c printing current settings"
.iX "c determining drive type"
.iX "c detecting a disk change"
.IP
.TP
\&\fR\&\f(CW-T\fR\ 
.TQ
\&\fR\&\f(CW--type\fR
Print out the drive name of a floppy device. This is used by the
\&\fR\&\f(CWMAKEFLOPPIES\fR script. The drive name is a letter (describing the
drive type) followed by the capacity of the format in bytes. The letter
is E for 3.5 ED drives, H for 3.5 HD drives, D for 3.5 DD drives, h for
5.25 HD drives and d for 5.25 DD drives. The drive type letter
corresponds to the oldest drive type supporting the format of this
device node (not necessarily the type of the drive referred by this
node.)  For the generic format nodes (/dev/fd0 et al.)  the name of
"native format" of the drive is printed, and for the default formats, if
a generic format has been redefined, its name becomes \fR\&\f(CW(null)\fR.
.TP
\&\fR\&\f(CW-p\fR\ 
.TQ
\&\fR\&\f(CW--print\fR
Prints out the configuration of the drive. The names of the various
fields are the same as the names of the option to set them, see below.
.TP
\&\fR\&\f(CW-P\fR\ 
.TQ
\&\fR\&\f(CW--printstate\fR
Prints out the cached internal state of the driver. The first line lists
various attributes about the disk:
.RS
.TP
\&\fR\&\f(CWdrive\ present\fR\ 
.TQ
\&\fR\&\f(CWdisk\ present\fR
.TQ
\&\fR\&\f(CWdisk\ writable\fR
These are only updated when the drive is accessed.
.TP
\&\fR\&\f(CWspinup\fR\ 
is the time when the motor became switched on for the last time.
.TP
\&\fR\&\f(CWselect\fR\ 
is the time when the drive became selected for the last time
.TP
\&\fR\&\f(CWfirst_read\fR\ 
is the time when the first read request after the last spin up
completed.
.TP
\&\fR\&\f(CWprobed_fmt\fR\ 
is the index of the autodetected format in the autodetection
sequence for this drive.
.TP
\&\fR\&\f(CWcylinder\fR\ 
is the cylinder where the drive head currently sits.  If this number is negative, it has the following meaning:
.RS
.TP
* \ \ 
-1 means that the driver doesn't know, but the controller does (a seek
command must be issued).
.TP
* \ \ 
-2 means that the controller doesn't know either, but is sure that it
not beyond the 80th track.  The drive needs a recalibration.
.TP
* \ \ 
-3 means that the head may be beyond the 80th track.  The drive needs
two successive recalibrations, because at each recalibration, the
controller only issues 80 move head commands per recalibration.
.RE
.TP
\&\fR\&\f(CWmaxblock\fR\ 
is the highest block number that has been read.
.TP
\&\fR\&\f(CWmaxcylinder\fR\ 
is a boolean which is set when a sector that is not on cylinder 0/head 0
has been read.  These are used for smart invalidation of the buffer
cache on geometry change.  The buffer cache of the drive is only
invalidated on geometry change when this change actually implies that a
block that has already been read changes position. This optimization is
useful for mtools which changes the geometry after reading the boot
sector.
.TP
\&\fR\&\f(CWgeneration\fR\ 
is roughly the number of disk changes noticed since boot. Disk changes
are noticed if the disk is actually changed, or if a flush command is
issued and for both cases if any I/O to/from the disk occurs. (i.e. if
you insert several disks, but don't do any I/O to them, the generation
number stays the same.)
.TP
\&\fR\&\f(CWrefs\fR\ 
is number of open file descriptors for this drive. It is always at
least one, because floppycontrol's file descriptor is counted too.
.TP
\&\fR\&\f(CWdevice\fR\ 
is format type (as derived from the minor device number) which is
currently being used.
.TP
\&\fR\&\f(CWlast_checked\fR\ 
is date (in jiffies) when the drive was last checked for a disk
change, and a disk was actually in the drive.
.RE
.IP
.TP
\&\fR\&\f(CW--pollstate\fR\ 
Polls the drive and then prints out the internal state of the
driver.(\fR\&\f(CW--Printstate\fR only prints out the cached information
without actually polling the drive for a disk change.)
.TP
\&\fR\&\f(CW--printfdcstate\fR\ 
Prints out the state of the controller where the target drive is
attached to.
.RS
.TP
\&\fR\&\f(CWspec1\fR\ 
.TQ
\&\fR\&\f(CWspec2\fR
are the current values of those registers.
.TP
\&\fR\&\f(CWrate\fR\ 
is current data transfer rate
.TP
\&\fR\&\f(CWrawcmd\fR\ 
is true if a raw command has been executed since the last reset. If this
is the case, a reset will be triggered when a drive on the same FDC is
next opened.
.TP
\&\fR\&\f(CWdor\fR\ 
is the value of the digital output register. The 4 high bits are a bit
mask describing which drives are spinning, the 2 low bits describe the
selected drive, bit 2 is used to reset the FDC, and bit 3 describes
whether this FDC has hold of the interrupt and the DMA. If you have two
FDCs, bit 3 is only set on one of them.
.TP
\&\fR\&\f(CWversion\fR\ 
is the version of the FDC. See \fR\&\f(CW\(iflinux/include/linux/fdreg.h\(is\fR for a
listing of the FDC version numbers.
.TP
\&\fR\&\f(CWreset\fR\ 
is true if a reset needs to be issued to the FDC before processing the
next request.
.TP
\&\fR\&\f(CWneed_configure\fR\ 
is true if this FDC needs configuration by the \fR\&\f(CWFD_CONFIGURE\fR
command.
.TP
\&\fR\&\f(CWhas_fifo\fR\ 
is set if the FDC understands the \fR\&\f(CWFD_CONFIGURE\fR command.
.TP
\&\fR\&\f(CWperp_mode\fR\ 
describes the perpendicular mode of this FDC. 0 is non-perpendicular mode,
2 is HD perpendicular mode, 3 is ED perpendicular mode, and 1 is unknown.
.TP
\&\fR\&\f(CWaddress\fR\ 
is the address of the first I/O port of the FDC.  Normally, this is
0x3f0 for the first FDC and 0x370 for the second.
.RE
.PP
.SH Drive\ type\ configuration\ and\ autodetection
.iX "c CMOS type"
.iX "c drive type"
.iX "c swapping drives"
.PP
The following options handle the different available drive types, such
as double density vs. high density vs. extra density drives, and 5 1/4
drives vs 3 1/2 drives.  Usually the drive type is stored in a
non-volatile memory, called CMOS, under the form of an integer ranging
from 1 to 6.
.PP
Different drive types are able to handle and autodetect different
formats (different autodetection lists). They also have different
"native format name". The native format is the "usual" format with the
highest capacity supported by the drive. (For example 720KB on a double
density 3 1/2 drive, and 1.2MB on a high density 5 1/4 drive.)
.PP
These settings are only changeable by the super user.
.TP
\&\fR\&\f(CW-c\ \fIcmos-type\fR\&\f(CW\fR\ 
.TQ
\&\fR\&\f(CW--cmos\ \fI\ cmos-type\fR\&\f(CW\fR
Set the virtual CMOS type of the floppy drive. This is useful if
.RS
.TP
* \ \ 
the physical CMOS type is wrong (this may happen with BIOSes
which use a non-standard mapping),
.TP
* \ \ 
you have more than two drives
(the physical CMOS may only describe up to two drives).
.TP
* \ \ 
you have a BIOS that allows swapping drives A: and B: for DOS.
.RE
 
Right now, this CMOS parameter is not used by the kernel, except for
feeding it back to other applications (for instance \fR\&\f(CWsuperformat\fR,
\&\fR\&\f(CWfloppymeter\fR or \fR\&\f(CWMAKEFLOPPIES\fR).  It is also possible to
supply a virtual CMOS type with the \fR\&\f(CWcmos\fR boot option
(see section  Boottime configuration).  If possible, I recommend you use the
boot option, rather than \fR\&\f(CWfloppycontrol\fR, because the boot option
also sets any parameters derived from the CMOS type, such as the
autodetection list and the native format, whereas \fR\&\f(CWfloppycontrol\fR
does not.
.TP
\&\fR\&\f(CW-A\ \ \fIautodetect-seq\fR\&\f(CW\ \fR\ 
.TQ
\&\fR\&\f(CW--autodetect\ \fI\ autodetect-seq\fR\&\f(CW\fR
Set the autodetection sequence (see section  Autodetection) The autodetection
sequence is a comma-separated list of at most eight format
descriptors. Each format descriptor is a format number optionally
followed by the letter \fR\&\f(CWt\fR.  For drive 0, the format number is the
minor device number divided by 4.  The autodetection sequence is used by
the driver to find out the format of a newly inserted disk. The formats
are tried one after the other, and the first matching format is
retained. To test the format, the driver tries to read the first sector
on the first track on the first head when \fR\&\f(CWt\fR is not given, or the

Changelog
020203	Fix fdmount and superformat bugs
020203	Create /dev/fd0 before testing for existence of /dev/fd0
011002	Fix floppycontrol.c printf conditional on FD_DISK_CHANGED, so that
	it works with gcc3 which apparently considers printf as a macro...
000612	Explicitly include sysmacros.h, apparently this is needed in
	certain environments
000603	Released 5.4
000527	Fixed a couple of buffer overrun exploits and other cleanup. Made
	fdutils compilable with Linux 2.3/2.4 kernels. Documentation fixes
970104	Fixed a few bugs listed in the Debian buglist posted to
	linux.debian.bugs.dis
	1.  Lessen dependancy of superformat on mtools by not failing if
	mformat cannot be called: low-level verify will still be done, but
	mbadblocks will be skipped
	2.  Fixed a few instances where the return value of getopt was
	misdeclared as a char instead of an integer
	3.  Fixed device names in fd.4
970411	Added error check to capacity measurement
	Corrected documentation Makefile bug
970629	Added vfat support to fdmount
970609	Fixed bug where a skew was erroneously applied to a 2m track
	Fixed typo in printfdprm
970604	Added vfat option for fdmount
970516	5.2 Released
970516	Fixed sunsite location in doc.
970430	Fixed typoes in doc about autodetection.
970428	Fixed manpage installation
970427	5.1 Released.  These repeated complaints about yylineno not
	defined start getting annoying ;-)
970426	More Makefile fixes.  Added note to doc about Alpha site
970422	Various Makefile fixes
970421	Lex yylineno fixes
970419	5.0 Released


src/floppycontrol.c
/*
 * floppycontrol.c
 *
 * This program illustrates the new FDGETEMSTRESH, FDSETMAXERRS, FDGETMAXERRS,
 * FDGETDRVTYP ioctl added by A. Knaff
 */

#include <stdio.h>
#include <sys/types.h>
#include <linux/fd.h>

#include <sys/ioctl.h>
#include <fcntl.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include "enh_options.h"

#ifndef FD_DISK_CHANGED
#define FD_DISK_CHANGED 0
#endif


int eioctl(int fd, int command,void * param, char *emsg)
{
  int r;
  if ((r=ioctl(fd,command,param))<0 )
    {
      perror(emsg);
      exit(1);
    }
  return r;
}

char *autodetect;
int reset_now;
struct floppy_drive_struct drivstat;
struct floppy_fdc_state fdcstat;
#ifdef FDWERRORGET
struct floppy_write_errors sta;
#endif
struct floppy_fdc_state fdcstat;
struct floppy_drive_params dpr;
#define max_error (dpr.max_errors)

int fd;

#define PRINT 0x1
#define SET_ERR 0x2
#define SET_EMSG 0x4
#define DO_FLUSH 0x8
#define DO_FMT_END 0x10
#define DO_DRVTYP 0x20
#define DO_EJECT 0x40
#define SET_DPR 0x80
#define PRINTSTATE 0x100
#define SET_AUTODETECT 0x200
#define POLLSTATE 0x400
#define SET_RESET 0x800
#define PRINTFDCSTATE 0x1000

#ifdef FDWERRORGET
#define CLRWERROR 0x2000
#define PRINTWERROR 0x4000
#endif

#ifndef FD_DEBUG
#define FD_DEBUG 2
#endif

struct enh_options optable[] = {
{ 'p', "print", 0, EO_TYPE_NONE, 0, PRINT, 0,
	  "print the current error thresholds and the drive type"},

{ 'a', "abort", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_ERR, 
	  (void *) &max_error.abort,
	  "set operation abortion threshold (if there are more errors than this threshold, the driver gives up to read/write the requested sector and reports an I/O error to the user program" },

{ 't', "readtrack", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_ERR, 
	  (void *) &max_error.read_track,
	  "set read track threshold (if there are less errors than this threshold, an entire track is read at once)" },

{ 'r', "recalibrate", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_ERR, 
	  (void *) &max_error.recal,
	  "set recalibration threshold (if there are more errors than this threshold, the driver recalibrates the drive, i.e. moves the head to track 0 and back, to make sure it really knows where the head is)" },

{ 'R', "reset", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_ERR, 
	  (void *) &max_error.reset,
	  "set reset threshold (if there are more errors than this threshold, the driver resets the floppy disk controller)" },

{ 'e', "reporting", 1, EO_TYPE_SHORT | EO_TYPE_DELAYED, 0, SET_ERR, 
	  (void *) &max_error.reporting,
	  "set error reporting threshold (if there are more errors than this threshold, error messages are printed to the console)" },

{ 'f', "flush", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, DO_FLUSH, 0,
	  "flush the floppy buffers" },

#ifdef FDEJECT
{ 'x', "eject", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, DO_EJECT, 0,
	  "eject the floppy disk" },
#endif

{ 'd', "drive", 1, EO_TYPE_FILE_OR_FD, 3, 0,
	  (void *) &fd,
	  "specify the target drive (default is /dev/fd0)" },

{ 'F', "formatend", 1, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, DO_FMT_END, 0,
	  "emits an FDFMTEND ioctl. This is needed when, after interrupting a formatting program, the drive light stays on" },

{ 'T', "type", 0, EO_TYPE_NONE, 0, DO_DRVTYP, 0,
	  "prints the drive type" },

{'\0', "debug", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, FD_DEBUG, SET_DPR,
	 (void *) &dpr.flags,
	 "switch debugging on" },

{'\0', "nodebug", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, ~FD_DEBUG, SET_DPR,
	 (void *) &dpr.flags,
	 "switch debugging off" },

{'\0', "message", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, FTD_MSG, SET_DPR,
	 (void *) &dpr.flags,
	 "switch autodetection and overrun messages on" },

{'\0', "nomessage", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, ~FTD_MSG,SET_DPR,
	 (void *) &dpr.flags,
	 "switch autodetection and overrun messages off" },

#ifdef FD_SILENT_DCL_CLEAR
{'\0', "silent_dcl_clear", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, 
	 FD_SILENT_DCL_CLEAR, SET_DPR,
	 (void *) &dpr.flags,
	 "switches on silent disk change status clear" },

{'\0', "noisy_dcl_clear", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED, 
	 ~FD_SILENT_DCL_CLEAR,SET_DPR,
	 (void *) &dpr.flags,
	 "switches off silent disk change status clear" },
#endif

{ 'c', "cmos", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.cmos,
	 "set cmos type" },

{ '\0', "hlt", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.hlt,
	 "set hlt" },

{ '\0', "hut", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.hut,
	 "set hut" },

{ '\0', "srt", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.srt,
	 "set srt" },

{ 'o', "spindown", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.spindown,
	  "set spindown time" },

{ 'u', "spinup", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.spinup,
	  "set spinup time" },

{ 's', "select_delay", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.select_delay,
	  "set drive select delay" },

{'\0', "rps", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.rps,
	 "sets rotations per second" },

{ 'O', "spindown_offset", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.spindown_offset,
	  "set spindown offset (decides in which position the disk stops)" },

{'\0', "cylinders", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.tracks,
	 "set maximal number of cylinders" },

{'\0', "tracks", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.tracks,
	 "obsolete (same as --cylinders)" },

{'\0', "timeout", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.timeout,
	 "set interrupt timeout" },

{ 'i', "interleave", 1, EO_TYPE_BYTE | EO_TYPE_DELAYED, 0, SET_DPR,
	 (void *) &dpr.interleave_sect,
	 "set interleave" },

{'A', "autodetect", 1, EO_TYPE_STRING, 0, SET_AUTODETECT,
	  (void *) &autodetect,
	  "set autodetection sequence (comma separated list of format indexes. The format index is the minor device number divided by four)" },

{ 'C', "checkfreq", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.checkfreq,
	  "set maximal disk change check interval" },

{ 'n', "native_format", 1, EO_TYPE_LONG | EO_TYPE_DELAYED, 0, SET_DPR,
	  (void *) &dpr.native_format,
	  "set the native format of this drive" },

{'\0', "resetnow", 1, EO_TYPE_LONG, 0, SET_RESET,
	 (void *) &reset_now,
	 "issue a reset command", },

{ 'P', "printstate", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, PRINTSTATE, 0,
	  "print internal per drive driver status"},

{ '\0', "printfdcstate", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, PRINTFDCSTATE, 0,
	  "print internal per fdc driver status"},

{ '\0', "pollstate", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, POLLSTATE, 0,
	  "polls internal driver status and prints it"},

#ifdef FDWERRORGET
{ '\0', "clrwerror", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, CLRWERROR, 0,
	  "clears write error structure"},

{ '\0', "printwerror", 0, EO_TYPE_NONE | EO_TYPE_DELAYED, 0, PRINTWERROR, 0,
	  "prints write error structure"},
#endif

#ifdef FD_BROKEN_DCL
{'\0', "broken_dcl", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,FD_BROKEN_DCL,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "work around broken disk change line" },
{'\0', "working_dcl", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,~FD_BROKEN_DCL,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "assume a working disk change line" },
#endif

#ifdef FD_INVERTED_DCL
{'\0', "inverted_dcl", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,
	 FD_INVERTED_DCL,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "assume an inverted disk change line" },
{'\0', "no_inverted_dcl", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,
	 ~FD_INVERTED_DCL,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "assume a non-inverted disk change line" },
#endif


{'\0', "messages", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,FTD_MSG,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "print informational messages" },
{'\0', "no_messages", 0, EO_TYPE_BITMASK_BYTE | EO_TYPE_DELAYED,~FTD_MSG,
	 SET_DPR,
	 (void *) &dpr.flags,
	 "don't print informational messages" },

{ '\0', 0 }
};

int main( int argc, char **argv)
{
	int ch;
	int mask=0;       
	char drivtyp[17];
	char *cont;
	int i;
		
	fd = -2;
	while((ch=getopt_enh(argc, argv, optable, 
			     0, &mask, "[drive]") ) != EOF ){
		if ( ch== '?' ){
			fprintf(stderr,"exiting\n");
			exit(1);
		}
		printf("unhandled option %d\n", ch);
		exit(1);
	}
	
	if ( fd == -1 )
		exit(0);

	if ( fd < 0 ){
		if ( optind < argc )
			fd = open(argv[optind], 3 | O_NDELAY);
		else
			fd = open("/dev/fd0", 3 | O_NDELAY);
		if ( fd < 0 ){
			perror("can't open floppy drive");
			print_usage_exit(argv[0],optable,"[drive]");
		}
	}
	
	if (mask & DO_FMT_END )
		eioctl(fd,FDFMTEND,0,"format end");

	if (mask & DO_FLUSH)
		eioctl(fd,FDFLUSH,0,"flush buffers");

#ifdef FDEJECT
	if (mask & DO_EJECT) {
		fsync(fd);
		eioctl(fd,FDEJECT,0,"eject floppy");
	}
#endif
	
	if (mask & (DO_DRVTYP ) )
		eioctl(fd,FDGETDRVTYP,(void *) drivtyp, "Get drive type");

	if ( mask & (PRINT | SET_DPR | SET_AUTODETECT ) )
		eioctl(fd,FDGETDRVPRM,(void *) &dpr, "Get drive parameters");
	else  if ( mask & SET_ERR )
		eioctl(fd,FDGETMAXERRS,(void *) &max_error, "Get max errors");
	
	if (mask & PRINT ){
		print_current_settings(optable, 
				       SET_ERR|SET_EMSG|SET_DPR);
		printf("autodetect seq.= ");
		for ( i=0; i<8; i++){
			if ( dpr.autodetect[i] == 0 )
				break;
			if ( i )
				putchar(',');
			printf("%1d", dpr.autodetect[i]);
			if ( dpr.read_track & ( 1 << i ) )
				putchar('t');
		}
		printf("\nflags: ");
		if(dpr.flags & 2 )
			printf("debug ");
		if(dpr.flags & FTD_MSG )
			printf("messages ");
#ifdef FD_BROKEN_DCL
		if(dpr.flags & FD_BROKEN_DCL )
			printf("broken_dcl");
#endif
#ifdef FD_SILENT_DCL_CLEAR
		if(dpr.flags & FD_SILENT_DCL_CLEAR)
			printf("silent_dcl_clear");
#endif
#ifdef FD_INVERTED_DCL
		if(dpr.flags & FD_INVERTED_DCL )
			printf("inverted_dcl");
#endif
		printf("\n");
	}

	if (mask & DO_DRVTYP )
		printf("%s\n", drivtyp);
	
	parse_delayed_options(optable, mask & (SET_ERR| SET_EMSG| SET_DPR));
	if ( mask & SET_AUTODETECT ){
		cont = autodetect;
		for ( i=0; i<8; i++){
			if ( *cont == '\0' )
				break;
			if ( *cont == ',' ){
				cont++;
				continue;
			}
			
			dpr.autodetect[i] = strtoul( cont, &cont, 0 );
			if ( *cont == 't' ){
				cont++;
				dpr.read_track |= 1 << i;
			} else
				dpr.read_track &= ~(1 << i);
			
			while ( *cont != ',' && *cont != '\0' )
				cont++;
			if ( *cont == ',' )
				cont++;
		}
		if ( i < 8 )
			dpr.autodetect[i] = 0;    
	}

	if (mask & SET_RESET)
		eioctl(fd, FDRESET, (void *)reset_now, "reset");

	if (mask & (PRINTSTATE | POLLSTATE) ){
		if ( mask & POLLSTATE )
			eioctl( fd, FDPOLLDRVSTAT, &drivstat,"get drive state");
		else
			eioctl( fd, FDGETDRVSTAT , &drivstat,"get drive state");

#ifndef FD_DCL_SEEN
# define FD_DCL_SEEN 0x40
#endif
		printf("%s %s %s %s %s %s\n", 
		       drivstat.flags & FD_VERIFY ? "verify" : "",
		       drivstat.flags & FD_DISK_NEWCHANGE ? "newchange" : "",
		       drivstat.flags & FD_NEED_TWADDLE ? "need_twaddle" : "",
		       drivstat.flags & FD_DISK_CHANGED ? "disk_changed" : "",
		       drivstat.flags & FD_DISK_WRITABLE ?"disk_writable" : "",
		       drivstat.flags & FD_DCL_SEEN ?"dcl_seen" : "");
		printf("spinup=		%ld\n", drivstat.spinup_date);
		printf("select=		%ld\n", drivstat.select_date);
		printf("first_read=	%ld\n", drivstat.first_read_date);
		printf("probed_fmt=    	%d\n", drivstat.probed_format);
		printf("cylinder=      	%d\n", drivstat.track);
		printf("maxblock=      	%d\n", drivstat.maxblock);
		printf("maxcylinder=	%d\n", drivstat.maxtrack);
		printf("generation=    	%d\n", drivstat.generation);
		printf("keep data=	%d\n", drivstat.keep_data);
		printf("refs=		%d\n", drivstat.fd_ref);
		printf("device=		%d\n", drivstat.fd_device);
		printf("last_checked=  	%ld\n", drivstat.last_checked);
	}


	if ( mask & PRINTFDCSTATE ){
		eioctl( fd, FDGETFDCSTAT , &fdcstat,"get fdc state");
		printf("spec1=%x\n", fdcstat.spec1);
		printf("spec2=%x\n", fdcstat.spec2);
		printf("rate=%x\n", fdcstat.dtr);
		printf("rawcmd=%x\n", fdcstat.rawcmd);
		printf("dor=%x\n", fdcstat.dor);
		printf("version=%x\n", fdcstat.version);
		printf("reset=%x\n", fdcstat.reset);
		printf("need_configure=%x\n", fdcstat.need_configure);
		printf("has_fifo=%x\n", fdcstat.has_fifo);
		printf("perp_mode=%x\n", fdcstat.perp_mode);
		printf("address=%x\n", (unsigned int) fdcstat.address);
	}

#ifdef FDWERRORGET
	if (mask & PRINTWERROR){
		eioctl( fd, FDWERRORGET , &sta,"get write error state");
		printf("write_errors %u  badness %u\n",
		       sta.write_errors, sta.badness);
		printf("first_error_sector %lu  first_error_generation %d\n",
		       sta.first_error_sector, sta.first_error_generation);
		printf("last_error_sector  %lu  last_error_generation  %d\n",
		       sta.last_error_sector, sta.last_error_generation);
	}
	if (mask & CLRWERROR)
		eioctl( fd, FDWERRORCLR , 0,"clear write error state");
#endif

	if (mask & ( SET_DPR | SET_AUTODETECT ))
		eioctl(fd,FDSETDRVPRM, (void *) &dpr, "Set drive parameters");
	else if (mask & SET_ERR )
		eioctl(fd,FDSETMAXERRS, (void *) &max_error, "Set max errors");

	exit(0);
}

